<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Estimate Effect Sizes, parameter estimates, etc.</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>

<table width="100%" summary="page for estimateEffectSizes {JunctionSeq}"><tr><td>estimateEffectSizes {JunctionSeq}</td><td align="right">R Documentation</td></tr></table>

<h2>
Estimate Effect Sizes, parameter estimates, etc.
</h2>

<h3>Description</h3>

<p>This function runs fits another generalized linear model to 
the data, this one intended for use in estimating the
effect sizes and expression estimates for each analysis.
</p>
<p>This function is called internally by the <code><a href="runJunctionSeqAnalyses.html">runJunctionSeqAnalyses</a></code> 
function, and thus for most purposes users should not need to call this
function directly. It may be useful to advanced users performing non-standard
analyses.
</p>


<h3>Usage</h3>

<pre>
   estimateEffectSizes(jscs, 
        method.expressionEstimation = c("feature-vs-gene","feature-vs-otherFeatures"),
        effect.formula = formula(~ condition + countbin + condition : countbin),
        geneLevel.formula = formula(~ condition),
        calculate.geneLevel.expression = TRUE,
        keep.estimation.fit = FALSE,
        nCores=1, 
        dispColumn="dispersion",
        verbose = TRUE)</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>jscs</code></td>
<td>

<p>A <code>JunctionSeqCountSet</code>. Usually initially created by 
<code><a href="readJunctionSeqCounts.html">readJunctionSeqCounts</a></code>. Size factors must be
set, usually using functions <code><a href="../../BiocGenerics/html/dge.html">estimateSizeFactors</a></code> and 
<code><a href="estimateJunctionSeqDispersions.html">estimateJunctionSeqDispersions</a></code>.
</p>
</td></tr>
<tr valign="top"><td><code>method.expressionEstimation</code></td>
<td>

<p>Character string. Can be used to apply alternative methodologies or implementations. Intended for advanced users who have strong opinions about the underlying statistical methodologies.
</p>
<p>Determines the methodology used to generate feature expression estimates and relative fold changes. By default each feature is modeled separately. Under the default count-vector method,
this means that the resultant relative fold changes will be a measure of the relative fold change between the feature and the gene as a whole.
</p>
<p>Alternatively, the &quot;feature-vs-otherFeatures&quot; method builds a large, complex model containing all features belonging to the gene. The coefficients for each feature
are then &quot;balanced&quot; using linear contrasts weighted by the inverse of their variance. In general we have found this method to produce very similar results but 
less efficiently and less consistently. Additionally, this alternative method &quot;multi-counts&quot; reads that cover more than one feature. This can result in over-weighting
of exonic regions with a large number of annotated variations in a small genomic area, as each individual read or read-pair may be counted many times in the model.
</p>
<p>Under the default option, no read or read-pair is ever counted more than once in a given model.
</p>
</td></tr>
<tr valign="top"><td><code>effect.formula</code></td>
<td>

<p>For advanced users. The base formula for the model used for effect size estimation.
</p>
<p>NOTE: the biological condition to be tested must be named &quot;condition&quot;.
</p>
</td></tr>
<tr valign="top"><td><code>geneLevel.formula</code></td>
<td>

<p>For advanced users. The base formula for the model used to estimate total gene-level expression.
</p>
<p>NOTE: the biological condition to be tested must be named &quot;condition&quot;.
</p>
</td></tr>
<tr valign="top"><td><code>calculate.geneLevel.expression</code></td>
<td>

<p>Logical value. If <code>TRUE</code>, gene-level expression will be estimated using the same maximum-likelihood method used in other analyses. Default: <code>TRUE</code>.
</p>
</td></tr>
<tr valign="top"><td><code>keep.estimation.fit</code></td>
<td>

<p>Logical value. If <code>TRUE</code>, save the complete model fits for every gene. This will require a lot of memory, but may be useful for statistical diagnostics. Default: <code>FALSE</code>.
</p>
</td></tr>
<tr valign="top"><td><code>nCores</code></td>
<td>

<p>The number of cores to use. Note that multicore functionality may not be available on all platforms. 
JunctionSeq attempts to use the BiocParallell package if it can be found installed. Otherwise it will
attempt to fallback to the multicore package. If neither package can be found it will fallback to single
core execution.
</p>
</td></tr>
<tr valign="top"><td><code>dispColumn</code></td>
<td>

<p>Character value. The name of the <code>fData(jscs)</code> column in which the model dispersion is stored.
</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>if TRUE, send debugging and progress messages to the console / stdout.</p>
</td></tr>
</table>


<h3>Value</h3>

<p>A JunctionSeqCountSet, with effect size results included.
</p>

<hr><div align="center">[Package <em>JunctionSeq</em> version 0.4.07 <a href="00Index.html">Index</a>]</div>
</body></html>
