<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: generateCompleteResults (DEPRECIATED)</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>

<table width="100%" summary="page for generateCompleteResults {JunctionSeq}"><tr><td>generateCompleteResults {JunctionSeq}</td><td align="right">R Documentation</td></tr></table>

<h2>
generateCompleteResults (DEPRECIATED)
</h2>

<h3>Description</h3>

<p>This function is DEPRECIATED. Use writeCompleteResults instead.
</p>
<p>This function takes the raw DEXSeq results and merges in feature 
annotations, as well as calculating and merging in a number of 
different normalized and fitted values for each level of the 
condition variable.
</p>


<h3>Usage</h3>

<pre>
generateCompleteResults(jscs, outfile.prefix = NULL, 
                        verbose = TRUE,nCores=1,
                        gzip.output = TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>jscs</code></td>
<td>

<p>A <code>JunctionSeqCountSet</code>. Usually created by <code><a href="runJunctionSeqAnalyses.html">runJunctionSeqAnalyses</a></code>.
</p>
<p>Alternatively, this can be created manually by <code><a href="readJunctionSeqCounts.html">readJunctionSeqCounts</a></code>. 
Dispersions and size factors must then be
set, usually using functions <code><a href="../../BiocGenerics/html/dge.html">estimateSizeFactors</a></code> and 
<code><a href="estimateJunctionSeqDispersions.html">estimateJunctionSeqDispersions</a></code>. Hypothesis tests must
be performed by <code><a href="testForDiffUsage.html">testForDiffUsage</a></code>. 
</p>
</td></tr>
<tr valign="top"><td><code>outfile.prefix</code></td>
<td>
<p>A string indicating the filename prefix where output files should be saved. If NULL, no data will be saved to file. (Default=NULL)</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>A boolean flag indicating whether or not to print progress information during execution. (Default=TRUE)</p>
</td></tr>
<tr valign="top"><td><code>nCores</code></td>
<td>
<p>If the package multicore is installed, the number of cores to use. If multicore is not installed, then setting this option anything other than the default of 1 does nothing but produce a warning.</p>
</td></tr>
<tr valign="top"><td><code>gzip.output</code></td>
<td>
<p>A boolean indicating whether output files should be gzipped.</p>
</td></tr>
</table>


<h3>Value</h3>

<p>A data frame containing the results for all listed splice sites of 
genes for which the GLM converged successfully. 
</p>
<p>Each row is a splice site. The columns are:
</p>

<ul>
<li><p> First, All columns from the res input file.
</p>
</li>
<li><p> exprVST_conditionVal: the variance-stabilizing-transformed fitted expression estimates for the splice site at each condition value.
</p>
</li>
<li><p> expr_conditionVal: the un-transformed fitted expression estimates for the splice site at each condition value.
</p>
</li>
<li><p> rExprVST_conditionVal: the variance-stabilizing-transformed fitted relative expression estimates for the splice site at each condition value.
</p>
</li>
<li><p> rExpr_conditionVal: as above, but without the Variance-stabilizing transformation.
</p>
</li>
<li><p> normCountVST_sampleID: Variance-stabilizing-transformed normalized counts for the splice site at each sample.
</p>
</li>
<li><p> normCount_sampleID: Variance-stabilizing-transformed normalized counts for the splice site at each sample.
</p>
</li>
<li><p> Finally, (if anno.file is non-null), the columns from the annotation file, matched to the given genes/splice sites.
</p>
</li></ul>

<p>Assuming the annotation file is the standard tab-delimited annotation file generated by prepare_annotation_with_splices.py, 
the final columns will be:
</p>

<ul>
<li><p> featureType: the type of the feature (for these analyses, always equal to either &ldquo;splice_site&rdquo; or &ldquo;novel_splice_site&rdquo;.
</p>
</li>
<li><p> chrom: The chromosome position.
</p>
</li>
<li><p> start
</p>
</li>
<li><p> end
</p>
</li>
<li><p> strand: &lsquo;+&rsquo; or &lsquo;-&rsquo;
</p>
</li>
<li><p> gene_id
</p>
</li>
<li><p> part_number: the 3-digit sub-feature ID
</p>
</li>
<li><p> transcripts: the list of transcripts, delimited by &lsquo;+&rsquo;
</p>
</li></ul>


<hr><div align="center">[Package <em>JunctionSeq</em> version 0.3.40 <a href="00Index.html">Index</a>]</div>
</body></html>
