<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Class '&quot;JunctionSeqCountSet&quot;'</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/8.3/styles/github.min.css">
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/8.3/highlight.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/8.3/languages/r.min.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
</head><body>

<table width="100%" summary="page for JunctionSeqCountSet-class {JunctionSeq}"><tr><td>JunctionSeqCountSet-class {JunctionSeq}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Class <code>"JunctionSeqCountSet"</code></h2>

<h3>Description</h3>

<p>A <code>JunctionSeqCountSet</code> is a container class that 
contains all information pertaining to a JunctionSeq analysis
and dataset. In general, these methods and slots will not be used
by the end-users. In general, JunctionSeqCountSet objects will be 
created by <code><a href="readJunctionSeqCounts.html">readJunctionSeqCounts</a></code> or 
<code><a href="runJunctionSeqAnalyses.html">runJunctionSeqAnalyses</a></code> and are to be
manipulated byhigh-level JunctionSeq functions such as 
<code><a href="estimateEffectSizes.html">estimateEffectSizes</a></code> or <code><a href="fitJunctionSeqDispersionFunction.html">fitJunctionSeqDispersionFunction</a></code>.
</p>
<p>The methods documented here are for use by advanced users only.
</p>


<h3>Details</h3>

<p>Slots:
</p>

<dl>
<dt><code>designColumns </code></dt><dd>  
<p>A character vector with the column names in the <code>design</code> data.frame.
</p>
</dd>
<dt><code>dispFitCoefs</code></dt><dd>  
<p>The dispersion fit coefficients.
</p>
</dd>
<dt><code>fittedMu</code></dt><dd>  
<p>Fitted mu values generated by DESeq2 code.
</p>
</dd>
<dt><code>dispFunctionType</code></dt><dd>  
<p>A list of various variables defining the dispersion functon used.
</p>
</dd>
<dt><code>dispFunction</code></dt><dd>  
<p>A function that converts a base mean to a fitted dispersion based on all included count bins.
</p>
</dd>
<dt><code>dispFunctionJct</code></dt><dd>  
<p>A function that converts a base mean to a fitted dispersion based only on the splice junction bins.
</p>
</dd>
<dt><code>dispFunctionExon </code></dt><dd>  
<p>A function that converts a base mean to a fitted dispersion based only on the exon bins.
</p>
</dd>
<dt><code>formulas</code></dt><dd>  
<p>A list of formulas used.
</p>
</dd>
<dt><code>annotationFile</code></dt><dd>  
<p>The annotation file.
</p>
</dd>
<dt><code>geneCountData</code></dt><dd>  
<p>A matrix of the gene-level counts
</p>
</dd>
<dt><code>countVectors</code></dt><dd>  
<p>A matrix of the count vectors.
</p>
</dd>
<dt><code>altSizeFactors</code></dt><dd>  
<p>(Not currently used)
</p>
</dd>
<dt><code>plottingEstimates</code></dt><dd>  
<p>A list of fitted estimates, for plotting.
</p>
</dd>
<dt><code>plottingEstimatesVST </code></dt><dd>  
<p>(Not currently used)
</p>
</dd>
<dt><code>geneLevelPlottingEstimates </code></dt><dd>  
<p>A list of gene-level fitted estimates, for plotting.
</p>
</dd>
<dt><code>modelFitForHypothesisTest </code></dt><dd>  
<p>(Not currently used)
</p>
</dd>
<dt><code>modelFitForEffectSize </code></dt><dd>  
<p>(Not currently used)
</p>
</dd>
<dt><code>flatGffData</code></dt><dd>  
<p>A <code>data.frame</code> representation of the flattened gff annotation for each countbin.
</p>
</dd>
<dt><code>flatGffGeneData</code></dt><dd>  
<p>A <code>data.frame</code> representation of the flattened gff annotation for each gene.
</p>
</dd>
<dt><code>analysisType</code></dt><dd>  
<p>The type of analysis. Character string.
</p>
</dd>
<dt><code>DESeqDataSet</code></dt><dd>  
<p>The specially-constructed <code>DESeqDataSet</code>, to be passed to the internally-loaded DESeq2 code.
</p>
</dd>
<dt><code>modelCoefficientsSample</code>:</dt><dd><p>Object of class <code>"list"</code>. Placeholder slot for model coefficients (used for diagnostic testing of code). </p>
</dd>
<dt><code>modelCoefficientsGene</code>:</dt><dd><p>Object of class <code>"list"</code>. Placeholder slot for model coefficients (used for diagnostic testing of code). </p>
</dd>
<dt><code>assayData</code>:</dt><dd><p>Object of class <code>"AssayData"</code>. Contains various data. </p>
</dd>
<dt><code>phenoData</code>:</dt><dd><p>Object of class <code>"AnnotatedDataFrame"</code>. Phenotype data. </p>
</dd>
<dt><code>featureData</code>:</dt><dd><p>Object of class <code>"AnnotatedDataFrame"</code>. Counting bin data. </p>
</dd>
<dt><code>experimentData</code>:</dt><dd><p>Object of class <code>"MIAxE"</code>. Information on the experiment. </p>
</dd>
<dt><code>annotation</code>:</dt><dd><p>Object of class <code>"character"</code>. Not used.</p>
</dd>
<dt><code>protocolData</code>:</dt><dd><p>Object of class <code>"AnnotatedDataFrame"</code>. Information on the code. </p>
</dd>
<dt><code>.__classVersion__</code>:</dt><dd><p>Object of class <code>"Versions"</code>. The version of the JunctionSeqCountSet. </p>
</dd>
</dl>



<h3>Constructor</h3>


<dl>
<dt></dt><dd>
<p><code>newJunctionSeqCountSet( countData, 
                              geneCountData, 
                              design, 
                              geneIDs, 
                              countbinIDs, 
                              featureIntervals=NULL, 
                              transcripts=NULL)</code>:
Creates a new JunctionSeqCountSet
</p>

<dl>
<dt><code>countData</code></dt><dd>
<p>A matrix of junction-level count data of non-negative integer values. 
The rows correspond to counts for each splice-junction counting bin, 
the columns correspond to samples. Note that biological 
replicates should each get their own column, while the 
counts of technical replicates (i.e., several sequencing 
runs/lanes from the same sample) should be summed up 
into a single column.
</p>
</dd>
<dt><code>geneCountData</code></dt><dd>
<p>A matrix of gene-level count data of non-negative integer values. 
The rows correspond to counts for each gene, 
the columns correspond to samples. Note that biological 
replicates should each get their own column, while the 
counts of technical replicates (i.e., several sequencing 
runs/lanes from the same sample) should be summed up 
into a single column. Must have the same dimensions as countData.
</p>
</dd>
<dt><code>design</code></dt><dd>
<p>A data frame consisting of all factors to be included in the analysis.
All columns should be factors. Each column should represent a different variable,
each row should represent a different sample. The number of rows must equal the
number of columns in geneCountData and countData.
</p>
</dd>
<dt><code>geneIDs</code></dt><dd>
<p>A character vector of gene indentifiers for each splice junction. The length must
equal the number of rows in countData.
</p>
</dd>
<dt><code>countbinIDs</code></dt><dd>
<p>A character vector of splice-junction-locus indentifiers for each splice junction. The length must
equal the number of rows in countData.
</p>
</dd>
<dt><code>featureIntervals</code></dt><dd>
<p>Optional. A data.frame with 4 columns: &quot;chr&quot;, &quot;start&quot;, &quot;end&quot;, and &quot;strand&quot;. chr and strand should 
be character vectors or factors, start and end must be integers.  
</p>
</dd>
<dt><code>transcripts</code></dt><dd>
<p>Optional. Character vector listing the transcripts that each splice junction belongs to. Some junctions
may belong to more than one transcripts. In this case, transcripts should be separated with the &quot;+&quot; character.
</p>
</dd>
</dl>

<p>This constructor function SHOULD NOT BE USED in normal operation. 
Instead you should use the <code><a href="readJunctionSeqCounts.html">readJunctionSeqCounts</a></code> 
function, which returns a new <code>JunctionSeqCountSet</code>.
</p>
</dd>
</dl>



<h3>Extends</h3>

<p>Class <code>"<a href="../../Biobase/html/class.eSet.html">eSet</a>"</code>, directly.
Class <code>"<a href="../../Biobase/html/class.VersionedBiobase.html">VersionedBiobase</a>"</code>, by class &quot;eSet&quot;, distance 2.
Class <code>"<a href="../../Biobase/html/class.Versioned.html">Versioned</a>"</code>, by class &quot;eSet&quot;, distance 3.
</p>


<h3>Note</h3>

<p>End-users generally will not use any of these slots or methods directly. 
However, they may be useful for model fit diagnostics and similar 
statistical experimentation.
</p>
<p>You can access method-mode information using the &quot;AltMethods&quot; attribute, and a list of all calls using the &quot;callStack&quot; attribute.
</p>


<h3>Author(s)</h3>

<p>Stephen Hartley
</p>


<h3>See Also</h3>

<p>The proper way to create a JunctionSeqCountSet is to use <code><a href="readJunctionSeqCounts.html">readJunctionSeqCounts</a></code> or 
<code><a href="runJunctionSeqAnalyses.html">runJunctionSeqAnalyses</a></code>.
</p>


<h3>Examples</h3>

<pre><code class="r">showClass(&quot;JunctionSeqCountSet&quot;)
</code></pre>

<pre><code>## Class &quot;JunctionSeqCountSet&quot; [package &quot;JunctionSeq&quot;]
## 
## Slots:
##                                                             
## Name:               designColumns               dispFitCoefs
## Class:                  character                    numeric
##                                                             
## Name:                    fittedMu           dispFunctionType
## Class:                     matrix                       list
##                                                             
## Name:                dispFunction            dispFunctionJct
## Class:                   function                   function
##                                                             
## Name:            dispFunctionExon                   formulas
## Class:                   function                       list
##                                                             
## Name:              annotationFile              geneCountData
## Class:                  character                     matrix
##                                                             
## Name:                countVectors             altSizeFactors
## Class:                     matrix                 data.frame
##                                                             
## Name:           plottingEstimates       plottingEstimatesVST
## Class:                       list                       list
##                                                             
## Name:  geneLevelPlottingEstimates  modelFitForHypothesisTest
## Class:                       list                       list
##                                                             
## Name:       modelFitForEffectSize                flatGffData
## Class:                       list                 data.frame
##                                                             
## Name:             flatGffGeneData               analysisType
## Class:                       list                  character
##                                                             
## Name:                DESeqDataSet    modelCoefficientsSample
## Class:               DESeqDataSet                       list
##                                                             
## Name:       modelCoefficientsGene                  assayData
## Class:                       list                  AssayData
##                                                             
## Name:                   phenoData                featureData
## Class:         AnnotatedDataFrame         AnnotatedDataFrame
##                                                             
## Name:              experimentData                 annotation
## Class:                      MIAxE                  character
##                                                             
## Name:                protocolData          .__classVersion__
## Class:         AnnotatedDataFrame                   Versions
## 
## Extends: 
## Class &quot;eSet&quot;, directly
## Class &quot;VersionedBiobase&quot;, by class &quot;eSet&quot;, distance 2
## Class &quot;Versioned&quot;, by class &quot;eSet&quot;, distance 3
</code></pre>


<hr /><div style="text-align: center;">[Package <em>JunctionSeq</em> version 1.5.4 <a href="00Index.html">Index</a>]</div>
</body></html>
