<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: JunctionSeqCountSet object and constructors</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>

<table width="100%" summary="page for JunctionSeqCountSet-class {JunctionSeq}"><tr><td>JunctionSeqCountSet-class {JunctionSeq}</td><td align="right">R Documentation</td></tr></table>

<h2>JunctionSeqCountSet object and constructors</h2>

<h3>Description</h3>

<p>A <code>JunctionSeqCountSet</code> is a container class that 
contains all information pertaining to a JunctionSeq analysis
and dataset.
</p>


<h3>Details</h3>

<p>Slots:
</p>

<dl>
<dt><code>designColumns </code></dt><dd>  
<p>A character vector with the column names in the <code>design</code> data.frame.
</p>
</dd>
<dt><code>dispFitCoefs</code></dt><dd>  
<p>The dispersion fit coefficients.
</p>
</dd>
<dt><code>fittedMu</code></dt><dd>  
<p>Fitted mu values generated by DESeq2 code.
</p>
</dd>
<dt><code>dispFunctionType</code></dt><dd>  
<p>A list of various variables defining the dispersion functon used.
</p>
</dd>
<dt><code>dispFunction</code></dt><dd>  
<p>A function that converts a base mean to a fitted dispersion based on all included count bins.
</p>
</dd>
<dt><code>dispFunctionJct</code></dt><dd>  
<p>A function that converts a base mean to a fitted dispersion based only on the splice junction bins.
</p>
</dd>
<dt><code>dispFunctionExon </code></dt><dd>  
<p>A function that converts a base mean to a fitted dispersion based only on the exon bins.
</p>
</dd>
<dt><code>formulas</code></dt><dd>  
<p>A list of formulas used.
</p>
</dd>
<dt><code>annotationFile</code></dt><dd>  
<p>The annotation file.
</p>
</dd>
<dt><code>geneCountData</code></dt><dd>  
<p>A matrix of the gene-level counts
</p>
</dd>
<dt><code>countVectors</code></dt><dd>  
<p>A matrix of the count vectors.
</p>
</dd>
<dt><code>altSizeFactors</code></dt><dd>  
<p>(Not currently used)
</p>
</dd>
<dt><code>plottingEstimates</code></dt><dd>  
<p>A list of fitted estimates, for plotting.
</p>
</dd>
<dt><code>plottingEstimatesVST </code></dt><dd>  
<p>(Not currently used)
</p>
</dd>
<dt><code>geneLevelPlottingEstimates </code></dt><dd>  
<p>A list of gene-level fitted estimates, for plotting.
</p>
</dd>
<dt><code>modelFitForHypothesisTest </code></dt><dd>  
<p>(Not currently used)
</p>
</dd>
<dt><code>modelFitForEffectSize </code></dt><dd>  
<p>(Not currently used)
</p>
</dd>
<dt><code>flatGffData</code></dt><dd>  
<p>A <code>data.frame</code> representation of the flattened gff annotation for each countbin.
</p>
</dd>
<dt><code>flatGffGeneData</code></dt><dd>  
<p>A <code>data.frame</code> representation of the flattened gff annotation for each gene.
</p>
</dd>
<dt><code>analysisType</code></dt><dd>  
<p>The type of analysis. Character string.
</p>
</dd>
<dt><code>DESeqDataSet</code></dt><dd>  
<p>The specially-constructed <code>DESeqDataSet</code>, to be passed to the internally-loaded DESeq2 code.
</p>
</dd>
</dl>



<h3>Constructor</h3>


<dl>
<dt></dt><dd>
<p><code>newJunctionSeqCountSet( countData, 
                              geneCountData, 
                              design, 
                              geneIDs, 
                              countbinIDs, 
                              featureIntervals=NULL, 
                              transcripts=NULL)</code>:
Creates a new JunctionSeqCountSet
</p>

<dl>
<dt><code>countData</code></dt><dd>
<p>A matrix of junction-level count data of non-negative integer values. 
The rows correspond to counts for each splice-junction counting bin, 
the columns correspond to samples. Note that biological 
replicates should each get their own column, while the 
counts of technical replicates (i.e., several sequencing 
runs/lanes from the same sample) should be summed up 
into a single column.
</p>
</dd>
<dt><code>geneCountData</code></dt><dd>
<p>A matrix of gene-level count data of non-negative integer values. 
The rows correspond to counts for each gene, 
the columns correspond to samples. Note that biological 
replicates should each get their own column, while the 
counts of technical replicates (i.e., several sequencing 
runs/lanes from the same sample) should be summed up 
into a single column. Must have the same dimensions as countData.
</p>
</dd>
<dt><code>design</code></dt><dd>
<p>A data frame consisting of all factors to be included in the analysis.
All columns should be factors. Each column should represent a different variable,
each row should represent a different sample. The number of rows must equal the
number of columns in geneCountData and countData.
</p>
</dd>
<dt><code>geneIDs</code></dt><dd>
<p>A character vector of gene indentifiers for each splice junction. The length must
equal the number of rows in countData.
</p>
</dd>
<dt><code>countbinIDs</code></dt><dd>
<p>A character vector of splice-junction-locus indentifiers for each splice junction. The length must
equal the number of rows in countData.
</p>
</dd>
<dt><code>featureIntervals</code></dt><dd>
<p>Optional. A data.frame with 4 columns: &quot;chr&quot;, &quot;start&quot;, &quot;end&quot;, and &quot;strand&quot;. chr and strand should 
be character vectors or factors, start and end must be integers.  
</p>
</dd>
<dt><code>transcripts</code></dt><dd>
<p>Optional. Character vector listing the transcripts that each splice junction belongs to. Some junctions
may belong to more than one transcripts. In this case, transcripts should be separated with the &quot;+&quot; character.
</p>
</dd>
</dl>

<p>This constructor function SHOULD NOT BE USED in normal operation. Instead you should use the <code><a href="readJunctionSeqCounts.html">readJunctionSeqCounts</a></code> function, which returns a new <code>JunctionSeqCountSet</code>.
</p>
</dd>
</dl>



<h3>See Also</h3>

<p><code><a href="readJunctionSeqCounts.html">readJunctionSeqCounts</a></code>
</p>

<hr><div align="center">[Package <em>JunctionSeq</em> version 0.5.1 <a href="00Index.html">Index</a>]</div>
</body></html>
