<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Read junctionSeq annotation files produced by QoRTs.</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>

<table width="100%" summary="page for readAnnotationData {JunctionSeq}"><tr><td>readAnnotationData {JunctionSeq}</td><td align="right">R Documentation</td></tr></table>

<h2>
Read junctionSeq annotation files produced by QoRTs.
</h2>

<h3>Description</h3>

<p>This function reads the &quot;flattened&quot; gff annotation file created by QoRTs. This annotation file contains all the gene, transcript, exon, and junction ID's and their loci.
</p>


<h3>Usage</h3>

<pre>
   readAnnotationData(flat.gff.file)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>flat.gff.file</code></td>
<td>

<p>Character string. The filename of the &quot;flat&quot; gff annotation file. The file may be gzip-compressed. 
This &quot;flat&quot; gff file must be produced by the QoRTs jar utility using the makeFlatGtf or 
mergeNovelSplices functions (depending on whether inclusion of novel splice junctions is desired).
</p>
</td></tr>
</table>

<hr><div align="center">[Package <em>JunctionSeq</em> version 0.3.5 <a href="00Index.html">Index</a>]</div>
</body></html>
